### PR TITLE
184 implement adversarial training support

### PR DESCRIPTION
**Pull request by fabiobrau**
This pull request introduces adversarial training support for PyTorch models in the codebase. It adds a general interface for adversarial trainers, a PyTorch-specific implementation following Madry et al.'s method, and example usage and tests to validate the new functionality. The changes also refactor the base PyTorch trainer to better support extensibility.

**Adversarial training support and interfaces:**

* Added `BaseAdvTrainer`, an abstract base class for adversarial model trainers, defining a standard interface for adversarial training across model types.
* Implemented `AdversarialTrainer` for PyTorch models, supporting adversarial training using a specified attack and a configurable strategy for combining clean and adversarial examples (currently only "adv" mode is implemented).

**Refactoring and extensibility improvements:**

* Refactored `BasePyTorchTrainer` to separate training for a single epoch (`train_epoch`) from the full training loop (`train`), improving clarity and enabling easier extension for custom training strategies. [[1]](diffhunk://#diff-6ce225308c0403c7eeeb38b26c9295b433d33af8d7786e190dc49411be1596a4L38-R42) [[2]](diffhunk://#diff-6ce225308c0403c7eeeb38b26c9295b433d33af8d7786e190dc49411be1596a4L55-R84)

**Examples and tests:**

* Added an example script (`run_adversarial_training.py`) demonstrating how to use the new adversarial training pipeline with a CIFAR-10 model, including evaluation before and after training.
* Introduced a new test for `AdversarialTrainer` to validate training with adversarial examples using a mock model and loss.

### CODE REVIEW

File: src/secmlt/defenses/adv_training/pytorch/adversarial_trainer.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            )
+            adv_data = attack(classifier, dataloader, stream=True)
+
+            # Combine the original and adversarial dataloaders
+            combined_data = self.collect_data(dataloader, adv_data, combining_mode)
+
+            # Train the model using the combined dataloader
+            model = model.train()
+            model = self.train_epoch(model, combined_data)
+            if self._scheduler is not None:
```
[the remark this answers is not in the document]
Comment (fabiobrau) [reply src/secmlt/defenses/adv_training/pytorch/adversarial_trainer.py]: @copilot apply changes based on [this feedback](https://github.com/pralab/secml-torch/pull/196#discussion_r3058129177)

File: examples/run_adversarial_training.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+# Fixed input normalization, implemented as a BatchNorm layer with the
+# CIFAR-10 mean/std baked into its running statistics.
+bn = torch.nn.BatchNorm2d(3, affine=False)
+bn.running_mean = torch.tensor(MEAN)
+bn.running_var = torch.tensor([std**2 for std in STD])
+bn.eval()
+
+
+def _keep_frozen(mode=True):
```
Comment (maurapintor) [inline examples/run_adversarial_training.py]: @fabiobrau is the user supposed to do this? Can't we wrap it internally and make it a default if this is required every time there is a batchnorm layer?

**Review by maurapintor — APPROVED** [review]
Looks good now. Thanks @fabiobrau !
